### DISCUSSION TITLE
Support replacement PRs on all managers

### DISCUSSION THREAD

**Discussion by HonkingGoose** [renovatebot/renovate#24883] [Suggest an Idea]
### What would you like Renovate to be able to do?

Right now we only raise replacement PRs for `npm` dependencies.

I think we want to support all managers eventually.
This is a big issue to track support for all managers:

- [ ] bazel
- [ ] git-submodules
- [ ] gomod
- [ ] gradle
- [x] hermit
- [ ] homebrew
- [x] maven
- [ ] regex
- [ ] sbt

Existing issues:

- [x] #12623
- [ ] #12625
- [x] #12626
- [x] ~#13211~
- [ ] #25643

Once we support _all_ managers, we should update the `readme.md` in this repository, and the `src/index.md` file in the docs repository. Both have a sentence that mentions this issue (#14149).

### If you have any ideas on how this should be implemented, please tell us here.

Nope.

### Is this a feature you are interested in implementing yourself?

No

**Comment by viceice** [comment]
We should link the already existing manager specific issues .

**Comment by JamieMagee** [comment]
#13211 will cover ~75% of managers. I need to get back on that PR

**Comment by hoerup** [comment]
Would it be possible to expand this logic with a "replace to none" / "remove"

**Comment by JamieMagee** [comment]
> Would it be possible to expand this logic with a "replace to none" / "remove"

Maybe? But not with the current set of changes that are in PR right now

**Comment by JamieMagee** [comment]
Closed:

- #13211
- #17476

In progress:

- #17881
- #17882
- #17883

**Comment by JamieMagee** [comment]
I've updated the list in the issue body to reflect the currently supported managers.

**Reply by MPV** [reply]
Could we also add `sbt` in the issue body as a manger not yet supported? 

It's tracked in:
- https://github.com/renovatebot/renovate/issues/25643

**Comment by HonkingGoose** [comment]
@rarkins you moved this issue to a discussion, but we still link to the old issue (#14149) on the:

- [ ] Docs homepage
- [ ] Main repository `readme.md`

What do you want me to do?

1. Update links on docs homepage and `readme.md` to point to this _discussion_
2. Drop the reference completely
   > Get replacement PRs to migrate from a deprecated dependency to the community suggested replacement, works with most managers, see [issue 14149](https://github.com/renovatebot/renovate/issues/14149) for exceptions
3. Copy information from discussion into a Renovate docs page (maybe [Renovate docs, Replacement Presets](https://docs.renovatebot.com/presets-replacements/), and link to that
4. Something else

**Reply by HonkingGoose** [reply]
@rarkins I hope it's okay to poke you after about three months. 🙃 We're still referring to the previous issue, and this discussion is not resolved yet. So there's still things we need to do.

Can you decide what to do?

**Comment by TWiStErRob** [comment]
What would be required to get Maven replacements?
This would cover Java/Kotlin/Android/Groovy/Maven/Gradle/Bazel ecosystems.

---

_Note: https://jakewharton.com/java-interoperability-policy-for-major-version-updates/ which suggests that relocations should happen on every major release, and it is happening on some._

**Reply by rarkins** [reply]
Support needs to be added to `updateDependency()` for each manager. Any which doesn't have its own `updateDependency()` function shouldn't need any changes and already work

**Comment by lyonlai** [comment]
👋 for Hermit manager, although it has an`updateDependency` method, it doesn't seem to impact the replacement PR functionalities.

I have used and tested the `replacementName` and `replacementVersion` very early days right after I developed this manager. Mainly  to replace dependencies managed by Hermit channel to specific version so that Renovate takes over the update in the future.

A recent issue brought this page into my attention. `hermit` is marked as not supporting replacement PRs that guides our engineers to have gone down to an alternative path where replacement PR could apply. I have spent some time digging into the source and retried this replacement functionality with `hermit` manager and found it works as expected.  

To avoid future confusion, I would like to tick the box for the `hermit` manager here. Would you mind me send a PR to update the doc? 

If there's anything missing please let me know.

**Reply by lyonlai** [reply]
Turn out I made a mistake in the previous statement. Only `replacementVersion` works and there changes are required in Hermit manager to have make `replacementName` works. I've added [a PR](https://github.com/renovatebot/renovate/pull/27573) for it. Please take a look.

**Comment by jamietanna** [comment]
I've made a stab at the support for `gomod` in https://github.com/renovatebot/renovate/pull/29575

**Comment by dsegurag** [comment]
Hi!

I want to expose my case.

I have a private GitLab instance deployed and I need to catch the tags of one repo. In order to do that, I'm using a `customManagers` like:

```json
"customManagers": [
  {
    "customType": "regex",
    "fileMatch": "Makefile",
    "matchStrings": ["PACKAGE_VERSION\\s:=\\s(?<currentValue>\\d+\\.\\d+.\\d+)"],
    "registryUrlTemplate": "https://gitlab.example.com",
    "depNameTemplate": "59064",
    "datasourceTemplate": "gitlab-tags",
    "versioningTemplate": "semver"
  }
]
```

The problem I see here is that the `depNameTemplate` property has the internal ID of the project I want to grab the tags and as a result, the dependency name it's `59064` which is not really nice...

Seems that having the `replacementName` property available for `regex` would be nice in my case.

Happy to hear any other way to achieve it.

Thanks all!

**Reply by dsegurag** [reply]
Hi again,

I know that I can use the path instead of the project id (at least it's more readable than an id) on `depNameTemplate`, but would be nice to be able to rename the dependency for a better name. 😃

Thanks!

**Comment by nh2297** [comment]
@HonkingGoose could you update this issue to include that maven is supported.

See: https://github.com/renovatebot/renovate/releases/tag/39.111.0

**Reply by jamietanna** [reply]
Done :)

**Comment by jochenberger** [comment]
Are there any attempts to support this for Gradle yet? Or is that supposed to work already?

**Reply by jamietanna** [reply]
I don't see any PRs that have introduced it, so the top of this Discussion is still up-to-date - help appreciated!
